src/rgw/rgw_acl_swift.cc

```diff
@@ -64,99 +66,123 @@ static bool uid_is_public(const string& uid)
          sub.compare(".referrer") == 0;
 }
 
-static bool extract_referer_urlspec(const std::string& uid,
-                                    std::string& url_spec)
+static bool is_referrer(const std::string& designator)
 {
-  const size_t pos = uid.find(':');
-  if (string::npos == pos) {
-    return false;
-  }
-
-  const auto sub = uid.substr(0, pos);
-  url_spec = uid.substr(pos + 1);
-
-  return sub.compare(".r") == 0 ||
-         sub.compare(".ref") == 0 ||
-         sub.compare(".referer") == 0 ||
-         sub.compare(".referrer") == 0;
+  return designator.compare(".r") == 0 ||
+         designator.compare(".ref") == 0 ||
+         designator.compare(".referer") == 0 ||
+         designator.compare(".referrer") == 0;
 }
 
-static bool normalize_referer_urlspec(string& url_spec, bool& is_negative)
+static boost::optional<ACLGrant> referrer_to_grant(std::string url_spec,
+                                                   const uint32_t perm)
 {
+  /* This function takes url_spec as non-ref std::string because of the trim
+   * operation that is essential to preserve compliance with Swift. It con't
+   * be easily accomplished with boost::string_ref. */
   try {
+    bool is_negative;
+    ACLGrant grant;
+
     if ('-' == url_spec[0]) {
-      is_negative = true;
       url_spec = url_spec.substr(1);
+      boost::algorithm::trim_if(url_spec, boost::is_any_of(" \t\r\n"));
```

Comment:
@rzarzynski if these strings are only passed as headers, i don't think we should be trying to handle control characters like `\t\r\n` here. looking at http://docs.openstack.org/developer/swift/overview_acl.html:
> Spaces may occur between elements as shown in the following example:
> 
> `.r : *, .rlistings, 7ec59e87c6584c348b563254aae4c221:*`